src/pd/client.rs

```diff
@@ -11,247 +11,408 @@
+        }
 
-    // Send a HTTP header to tell PD to hijack this connection for RPC.
-    let header_str = format!("GET {} HTTP/1.0\r\n\r\n", PD_RPC_PREFIX);
-    let header = header_str.as_bytes();
-    match stream.write_all(header) {
-        Ok(_) => Ok(stream),
-        Err(err) => Err(box_err!("failed to connect to {} error: {:?}", endpoint, err)),
+        Err(_) => {
+            let mut parts = addr.split(':');
+            (parts.next().unwrap().to_owned(), parts.next().unwrap().parse::<u16>().unwrap())
+        }
+    };
+
+    let mut conf: grpc::client::GrpcClientConf = Default::default();
+    conf.http.no_delay = Some(true);
+
+    // TODO: It seems that `new` always return an Ok(_).
+    match pdpb_grpc::PDClient::new(&host, port, false, conf) {
+        Ok(client) => {
+            // try request.
+            match client.GetMembers(pdpb::GetMembersRequest::new()) {
+                Ok(_) => Ok(client),
+                Err(e) => Err(box_err!("fail to connect to {:?} {:?}", addr, e)),
+            }
+        }
+
+        Err(e) => Err(box_err!(e)),
     }
 }
 
-impl RpcClientCore {
-    fn new(endpoints: Vec<String>) -> RpcClientCore {
-        RpcClientCore {
-            endpoints: endpoints,
-            stream: None,
+fn try_connect_leader(members: &pdpb::GetMembersResponse)
+                      -> Result<(pdpb_grpc::PDClient, pdpb::GetMembersResponse)> {
+    // Try to connect the PD cluster leader.
+    let leader = members.get_leader();
+    for ep in leader.get_client_urls() {
+        if let Ok(client) = connect(ep.as_str()) {
+            info!("connect to PD leader {:?}", ep);
+            return Ok((client, members.clone()));
         }
     }
 
-    fn try_connect(&mut self) -> Result<()> {
-        // Randomize endpoints.
-        let len = self.endpoints.len();
-        let mut indexes: Vec<usize> = (0..len).collect();
-        rand::thread_rng().shuffle(&mut indexes);
-
-        for i in indexes {
-            let ep = &self.endpoints[i];
-            match rpc_connect(ep.as_str()) {
-                Ok(stream) => {
-                    info!("PD client connects to {}", ep);
-                    self.stream = Some(stream);
-                    return Ok(());
+    // Then try to connect other members.
+    // Randomize endpoints.
+    let members = members.get_members();
+    let mut indexes: Vec<usize> = (0..members.len()).collect();
+    rand::thread_rng().shuffle(&mut indexes);
+
+    for i in indexes {
+        for ep in members[i].get_client_urls() {
+            match connect(ep.as_str()) {
+                Ok(cli) => {
+                    let resp = match cli.GetMembers(pdpb::GetMembersRequest::new()) {
+                        Ok(resp) => resp,
+                        Err(e) => {
+                            error!("PD endpoint {} failed to respond: {:?}", ep, e);
+                            continue;
+                        }
+                    };
+
+                    return Ok((cli, resp));
```

Comment:
But the RpcClient is for not leader, so next call will fail and then try call leader again?